common/hyperscaler/rules/rule.go

```diff
@@ -0,0 +1,224 @@
+			return attr.Setter(r, value)
+		}
+	}
+
+	return nil, fmt.Errorf("unknown attribute %s", attribute)
+}
+
+func (r *Rule) NumberOfInputAttributes() int {
+	count := 0
+
+	for _, attr := range InputAttributes {
+		value := attr.Getter(r)
+		if value != "" {
+			count++
+		}
+	}
+
+	return count
+}
+
+func (r *Rule) String() string {
+	ruleStr := r.StringNoLabels()
+
+	labels := r.CalculateLabels()
+	labelsStr := "# "
+	labelsToSort := make([]string, 0, len(labels))
+	for key, value := range labels {
+		labelsToSort = append(labelsToSort, fmt.Sprintf("%s: %s", key, string(value)))
+	}
+
+	sort.Strings(labelsToSort)
+
+	for _, key := range labelsToSort {
+		labelsStr += fmt.Sprintf("%s, ", key)
+	}
+
+	labelsStr = labelsStr[:len(labelsStr)-2]
+
+	return fmt.Sprintf("%-50s %-50s", ruleStr, labelsStr)
+}
+
+func (r *Rule) StringNoLabels() string {
+	ruleStr := fmt.Sprintf("%s", r.Plan)
+
+	if r.ContainsInputAttributes {
+
+		ruleStr += fmt.Sprintf("(")
+
+		for _, attr := range InputAttributes {
+			attrStr := attr.String(r)
+			ruleStr += fmt.Sprintf("%s", attrStr)
+		}
+
+		ruleStr = ruleStr[:len(ruleStr)-2]
+		ruleStr += fmt.Sprintf(")")
+	}
+
+	if r.ContainsOutputAttributes {
+		ruleStr += fmt.Sprintf("-> ")
+
+		for _, attr := range OutputAttributes {
+			attrStr := attr.String(r)
+			ruleStr += fmt.Sprintf("%s", attrStr)
+		}
+
+		ruleStr = ruleStr[:len(ruleStr)-2]
+	}
+
+	return ruleStr
+}
+
+func (r *Rule) IsResolved() bool {
+	resolved := true
+
+	for _, attr := range InputAttributes {
+		value := attr.Getter(r)
+		resolved = resolved && value != ASTERISK
+	}
+
+	return resolved
+}
+
+func (r *Rule) Combine(rule Rule) *Rule {
+	newRule := NewRule()
+	newRule.SetPlan(r.Plan)
+
+	for _, attr := range InputAttributes {
+		value := attr.Getter(r)
+		if value != "" && value != ASTERISK {
+			attr.Setter(newRule, value)
+			newRule.ContainsInputAttributes = true
+		} else {
+			valueR := attr.Getter(&rule)
+			attr.Setter(newRule, valueR)
+			newRule.ContainsInputAttributes = true
+		}
+	}
+
+	newRule.hyperscalerNameMappingFunction = r.hyperscalerNameMappingFunction
+
+	return newRule
+}
+
+func (r *Rule) SignatureWithValues() string {
+	key := r.Plan
+
+	for _, attr := range InputAttributes {
+		key += attr.Name + SIGNATURE_ATTR_SEPARATOR
+		checkValue := attr.Getter(r)
+		key += getAttrValueSymbol(checkValue, ASTERISK, checkValue)
+	}
+
+	return key
+}
+
+func (r *Rule) MirroredSignature() string {
+	return r.SignaturePreKeys(ATTRIBUTE_WITH_VALUE, ASTERISK)
+}
+
+func (r *Rule) SignaturePreKeys(positiveKey, mirroredKey string) string {
+	signatureKey := r.Plan
+
+	for _, attr := range InputAttributes {
+		signatureKey += attr.Name + SIGNATURE_ATTR_SEPARATOR
+		checkValue := attr.Getter(r)
+		signatureKey += getAttrValueSymbol(checkValue, positiveKey, mirroredKey)
+	}
+
+	return signatureKey
+}
+
+func getAttrValueSymbol(checkedValue, returnedValueTrue, returnedValueFalse string) string {
+	if checkedValue == "" || checkedValue == ASTERISK {
```

Comment:
And again we check equivalent of:
`if value != "" && value != ASTERISK {`
but this time negated.